datafusion/functions/src/unicode/rpad.rs

```diff
@@ -77,96 +82,122 @@ impl ScalarUDFImpl for RPadFunc {
+                    if length == 0 {
+                        Ok(Some("".to_string()))
+                    } else {
+                        let graphemes = string.graphemes(true).collect::<Vec<&str>>();
+                        if length < graphemes.len() {
+                            Ok(Some(graphemes[..length].concat()))
                         } else {
-                            let graphemes = string.graphemes(true).collect::<Vec<&str>>();
-                            if length < graphemes.len() {
-                                Ok(Some(graphemes[..length].concat()))
+                            let mut s = string.to_string();
+                            if !$is_view {
+                                s.push_str(&" ".repeat(length - graphemes.len()));
                             } else {
-                                let mut s = string.to_string();
                                 s.push_str(" ".repeat(length - graphemes.len()).as_str());
-                                Ok(Some(s))
                             }
+                            Ok(Some(s))
                         }
                     }
-                    _ => Ok(None),
-                })
-                .collect::<Result<GenericStringArray<T>>>()?;
+                }
+                _ => Ok(None),
+            })
+            .collect::<Result<GenericStringArray<T>>>()
+    }};
+
+    // For the three-argument case
+    ($string_array:expr, $length_array:expr, $fill_array:expr, $is_view:expr) => {{
+        $string_array
+            .iter()
+            .zip($length_array.iter())
+            .zip($fill_array.iter())
+            .map(|((string, length), fill)| match (string, length, fill) {
+                (Some(string), Some(length), Some(fill)) => {
+                    if length > i32::MAX as i64 {
+                        return exec_err!("rpad requested length {} too large", length);
+                    }
+
+                    let length = if length < 0 { 0 } else { length as usize };
+                    let graphemes = string.graphemes(true).collect::<Vec<&str>>();
+                    let fill_chars = fill.chars().collect::<Vec<char>>();
+
+                    if length < graphemes.len() {
+                        Ok(Some(graphemes[..length].concat()))
+                    } else if fill_chars.is_empty() {
+                        Ok(Some(string.to_string()))
+                    } else {
+                        let mut s = string.to_string();
+                        let char_vector: Vec<char> = (0..length - graphemes.len())
+                            .map(|l| fill_chars[l % fill_chars.len()])
+                            .collect();
+                        s.push_str(&char_vector.iter().collect::<String>());
+                        Ok(Some(s))
+                    }
+                }
+                _ => Ok(None),
+            })
+            .collect::<Result<GenericStringArray<T>>>()
```

Comment:
This means the result of `rpad` is always `GenericStringArray` (`Utf8Array`) right?

I wonder if it would make sense to produce the same output type as the first argument 🤔  For the string view case eventually I could imagine reusing the same underlying buffers if the string had nothing to pad.